### ISSUE TITLE
Alter the Drag and Drop command

### ISSUE THREAD

**Issue by j718** [equalsraf/neovim-qt#659]
I am trying to insert images into my markdown by drag and drop. So far I have the following script to save them to my default image location and then copy them to the register. Is there any way to edit the drag and drop to handle jpg files without opening a buffer for them?

```
fu! Imgsave()
  let filename=expand("%:t")
  let img_name="images/" . localtime() . filename
  cd ~/repos/notes
  execute "silent w" img_name
  call setreg("", ["", "![" . filename . "]" . "(../" .img_name . ")"])
  History
endfunction

au BufReadPost,BufNewFile *.png,*.jpeg,*.jpg :call Imgsave()
```

The main issue with this approach is that I have to close the new buffer an navigate back to the most recent buffer. It would nice to just not open a new buffer to begin with.

**Comment by jgehrig** [comment]
@j718

Yes, I think the behavior you are looking for is possible. The Drag and Drop feature is simply a call to `function GuiDrop(...)` in `src/gui/runtime/plugin/nvim_gui_shim.vim`.

If you can override that function, you have an entry point to handle Drag and Drop events however you'd like with vimscript.

Maybe we can add something to make overriding the function easier:
```patch
--- a/src/gui/runtime/plugin/nvim_gui_shim.vim  Sat Mar 28 02:54:55 2020 -0400
+++ b/src/gui/runtime/plugin/nvim_gui_shim.vim  Sun Mar 29 14:01:00 2020 -0400
@@ -98,7 +98,11 @@
 
 " GuiDrop('file1', 'file2', ...) is similar to :drop file1 file2 ...
 " but it calls fnameescape() over all arguments
-function GuiDrop(...)
+function! GuiDrop(...)
+       if exists('*GuiDropCustom')
+               return call("GuiDropCustom", a:000)
+       endif
+
        let l:fnames = deepcopy(a:000)
        let l:args = map(l:fnames, 'fnameescape(v:val)')
        exec 'drop '.join(l:args, ' ')
```

vimrc example:
```
" This is just an example, it appends all dropped filenames into the current buffer.
function! GuiDropCustom(...)
	let l:fnames = deepcopy(a:000)
	let l:args = map(l:fnames, 'fnameescape(v:val)')

	for l:file in l:args
		echo l:file
		"let line = getline('.')
		"call setline('.', strpart(line, 0, col('.') - 1) . text . strpart(line, col('.') - 1))
		call append(line('.'), l:file)
	endfor
endfunction
```

**Comment by j718** [comment]
Yeah that would be perfect!

**Comment by jgehrig** [comment]
Could you test this feature and share some sample code for your use case?

It would be nice to have an example of how use `:GuiDropCustom` in the Wiki along with some sample code.

**Comment by j718** [comment]
I haven't been able to test it yet, b/c I don't know how to download and build the binaries from your pull request. I will add them once it is merged into the nightly builds.

If you have input on how to use your pull request, I can do it earlier. Thanks again.

**Comment by jgehrig** [comment]
@j718 No worries.

If you are running Windows, you can find the build here: [AppVeyor Artifacts](https://ci.appveyor.com/project/equalsraf/neovim-qt/builds/31947337/job/449b6bu05v60v0ad/artifacts)

The changes are in the vimscript runtime, so you can easily modify this in-place without building. The latest versions have a `--version` argument that will make finding your runtime directory easier.

Run `nvim-qt --version`:
```
NVIM-QT v0.2.14.0
Build type: Release
Compilation:-march=native -O2 -pipe -Wall -Wextra -Wno-unused-parameter -std=c++11 -Wfatal-errors
Qt Version: 5.13.2
Environment: 
  nvim: nvim
  args: --cmd let &rtp.=',/usr/share/nvim-qt/runtime' --cmd set termguicolors
  runtime: /usr/share/nvim-qt/runtime
...
```

You can see my runtime is `runtime: /usr/share/nvim-qt/runtime`.

You need to do is add the new lines to `plugin/nvim_gui_shim.vim`:
```
" Check for user defined handler function
if exists('*GuiDropCustomHandler')
	return call("GuiDropCustomHandler", a:000)
endif
```
Note, the custom handler name will likely be **GuiDropCustomHandler** in the final release.

**Comment by jgehrig** [comment]
FYI @j718

These changes are now be available in the latest nightly release of neovim.

https://github.com/equalsraf/neovim-qt/wiki/Configuration-Options#guidropcustomhandler

Let me know if you settle on a configuration for this feature. It would be great to update the Wiki with a real-world example.
